src/pool.jl

```diff
@@ -225,33 +161,36 @@ function Base.append!(pool::CategoricalPool, levels)
     return pool
 end
 
-function Base.delete!(pool::CategoricalPool{S}, levels...) where S
-    for level in levels
-        levelS = convert(S, level)
-        if haskey(pool.invindex, levelS)
-            ind = pool.invindex[levelS]
-            delete!(pool.invindex, levelS)
-            splice!(pool.index, ind)
-            ord = splice!(pool.order, ind)
-            splice!(pool.levels, ord)
-            splice!(pool.valindex, ind)
-            for i in ind:length(pool)
-                pool.invindex[pool.index[i]] -= 1
-                pool.valindex[i] = CategoricalValue(i, pool)
-            end
-            for i in 1:length(pool)
-                pool.order[i] > ord && (pool.order[i] -= 1)
-            end
-        end
+# Do not override Base.merge as for internal use we need to use the type and orderedness
+# of the first pool rather than promoting both pools
+function merge_pools(a::CategoricalPool{T, R}, b::CategoricalPool) where {T, R}
+    if length(a) == 0 && length(b) == 0
+        newlevs = T[]
+        ordered = isordered(a)
+    elseif length(a) == 0
+        newlevs = Vector{T}(levels(b))
+        ordered = isordered(b)
+    elseif length(b) == 0
+        newlevs = copy(levels(a))
+        ordered = isordered(a)
+    else
+        nl, ordered = mergelevels(isordered(a), a.levels, b.levels)
```

Comment:
I wonder whether this would better be
```suggestion
        nl, ordered = mergelevels(isordered(a) | isordered(b), a.levels, b.levels)
```
It could be a bit surprising that an array becomes ordered after setting a value. OTOH it's a bit weird too that we set the levels, but not orderedness, which introduces a small inconsistency with `vcat`. Opinions?

Reply:
So my thinking is the following. We have two use-cases for `merge_pools` (maybe we need an argument to differentiate between them):
* asymmetric - you merge `b` into `a` creating a new pool that is to replace `a` (setting a value in an array)
* symmetric - you merge `a` and `b` to create something new (like in `vcat`)

In these two scenarios I think we can have a different rules.

Now as for the rules - I am not sure what is best.

For symmetric (`vcat`) case the rule I think should be "treat unordered categorical vectors as if they were not categorical (but just plain vectors)", what I mean is that it should matter (i.e. the result should be the same) if I merge `["a", "b", "c"]` and `categorical(["a", "b", "c"])` with some other categorical array that is ordered.

For asymmetric case I think the right rule is to keep the status of the destination array (i.e. the one we write to).

Reply:
Actually `merge_pools` is only used for the asymmetric case. The symmetric case only happens with `vcat`, which uses its own logic.

> For symmetric (`vcat`) case the rule I think should be "treat unordered categorical vectors as if they were not categorical (but just plain vectors)", what I mean is that it should matter (i.e. the result should be the same) if I merge `["a", "b", "c"]` and `categorical(["a", "b", "c"])` with some other categorical array that is ordered.

Well that's already the case for the example you give since levels are the same, but that's quite different when arrays have unused levels (like we do with `setindex!`). Are you suggesting to change that?

> For asymmetric case I think the right rule is to keep the status of the destination array (i.e. the one we write to).

OK, so basically what we do now.

Reply:
> Are you suggesting to change that?

No (I did not consider unused levels when writing this - good point; I concentrated on orderness).

I understand that we propagate unused levels now - right? I think it is OK, as then the levels in the new pool are a superset for levels in pools from source arrays.